### PR TITLE
Validate frames conversions

### PR DESCRIPTION
**Pull request by jorgepiloto**
Tries to solve for #22. Although both APIs (orekit and poliastro ones) provide simple routines to convert between fixed and inertial frames for each celestial body, results diverge. I also tried to use Gmat as an extra checker, but output values were also different.
Still working on this... :factory:

### CODE REVIEW

**Comment by jorgepiloto** [comment]
Orekit implementation IUA poles: https://gitlab.orekit.org/orekit/orekit/-/blob/develop/src/main/java/org/orekit/bodies/PredefinedIAUPoles.java

**Comment by astrojuanlu** [comment]
I'm back! Let me know if I can help in any way 🚀

**Comment by jorgepiloto** [comment]
*Gmat definitions*

* http://gmat.sourceforge.net/docs/R2018a/html/CelestialBody.html#CelestialBody_Remarks

**Comment by jorgepiloto** [comment]
It seems that Orekit makes use of the [IAU WGCCRE 2009 report](https://astropedia.astrogeology.usgs.gov/download/Docs/WGCCRE/WGCCRE2009reprint.pdf) definitions instead of using [IAU WGCCRE 2015 report](https://astropedia.astrogeology.usgs.gov/download/Docs/WGCCRE/WGCCRE2015reprint.pdf) ones. This introduces numerical differences in Mercury, Mars and Neptune.

**Comment by jorgepiloto** [comment]
It seems that Orekit makes use of the [IAU WGCCRE 2009 report](https://astropedia.astrogeology.usgs.gov/download/Docs/WGCCRE/WGCCRE2009reprint.pdf) definitions instead of using [IAU WGCCRE 2015 report](https://astropedia.astrogeology.usgs.gov/download/Docs/WGCCRE/WGCCRE2015reprint.pdf) ones. This introduces numerical differences in Mercury, Mars and Neptune.

**Comment by jorgepiloto** [comment]
In fact, the [IAU report 2015](https://astropedia.astrogeology.usgs.gov/download/Docs/WGCCRE/WGCCRE2015reprint.pdf) for the case of Martian planet contains a more detailed numerical description than the one from the [IAU report 2009](https://astropedia.astrogeology.usgs.gov/download/Docs/WGCCRE/WGCCRE2009reprint.pdf) one...

**Comment by jorgepiloto** [comment]
Cool, all tests passing finally! 🚀

File: orekit/tests/validate_frames.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+
+@pytest.mark.parametrize("r_vec", R_SET)
+@pytest.mark.parametrize("v_vec", V_SET)
+@pytest.mark.parametrize("body_name", BODIES_NAMES)
+def validate_from_body_intertial_to_body_fixed(body_name, r_vec, v_vec):
+
+    # Conversion frames for these bodies are expected to fail since Orekit 10.3
+    # makes use of IAU WGCCRE 2009 report while poliastro uses IAU WGCCRE 2015
+    # report
+    if body_name in ["Mercury", "Mars", "Neptune"]:
```
Comment (astrojuanlu) [inline orekit/tests/validate_frames.py:153]: Let's parametrize this and use `pytest.param` instead, see https://github.com/poliastro/poliastro/search?q=pytest.param

**Comment by jorgepiloto** [comment]
Parametrize GMAT scripts using Jinja2: add a template script plus the generator one.

**Comment by astrojuanlu** [comment]
For the GMAT scripts, @jorgepiloto commented that they're 27 ** 2 😅  so let's use https://jinja.palletsprojects.com/en/2.11.x/ instead, and a small script to generate the scripts from the template.

File: gmat/tests/build_frames_scripts.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
+from itertools import product
```
Comment (astrojuanlu) [inline gmat/tests/build_frames_scripts.py:1]: Awesome!!! 👏🏽

File: gmat/tests/template_frames.script [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+%----------------------------------------
+%---------- Spacecraft
+%----------------------------------------
+
+% Build spacecraft's orbit from vectors
+Create Spacecraft ss_0;
+GMAT ss_0.DateFormat = TTGregorian;
+GMAT ss_0.CoordinateSystem = bodyICRF;
+GMAT ss_0.DisplayStateType = Cartesian;
+GMAT ss_0.X = {{ rx }};
```
Comment (jorgepiloto) [inline gmat/tests/template_frames.script:35]: The values used in here are any combination produced by {-1, 0, 1}. However, GMAT shows to perform very slowly with those although it finally converges.

Comment (astrojuanlu) [reply gmat/tests/template_frames.script:35]: What units are these? Meters?

Comment (astrojuanlu) [reply gmat/tests/template_frames.script:35]: Either way, we will use planet radiuses

**Comment by jorgepiloto** [comment]
Finally, this is ready for review @astrojuanlu 🚀

File: gmat/tests/build_frames_scripts.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    Uranus_poliastro,
+    Neptune_poliastro,
+]
+
+BODY_FROM_NAME = dict(zip(BODIES_NAMES, POLIASTRO_BODIES))
+
+if __name__ == "__main__":
+
+    # Open the template
+    with open("template_frames.script") as template_file:
```
Comment (astrojuanlu) [inline gmat/tests/build_frames_scripts.py:49]: What about moving this to a `main` function and do

```python
def main():
    ...


if __name__ == "__main__":
    main()
```

?

File: gmat/tests/build_frames_scripts.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        template_raw = Template(template_file.read())
```
Comment (astrojuanlu) [inline gmat/tests/build_frames_scripts.py:50]: After we read `template_file`, do we need all the rest to be inside the `with open(...)` block? I think we can move it outside

File: gmat/tests/build_frames_scripts.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+                    (rx, ry, rz), (vx, vy, vz) = r_vec, v_vec
+                    filename = f"gmat_validate_{body_name}_frames_R{rx}{ry}{rz}_V{vx}{vy}{vz}.script"
+
+                    # Compute Body radius in kilometers
+                    r_norm, v_norm = [norm(vec) for vec in [r_vec, v_vec]]
+                    R = BODY_FROM_NAME[body_name].R.to(u.km).value
+
+                    # Correction factor to normalize position and velocity vectors
+                    k_r = R / r_norm if r_norm != 0 else 1.00
+                    k_v = 1 / v_norm if v_norm != 0 else 1.00
```
Comment (astrojuanlu) [inline gmat/tests/build_frames_scripts.py:67]: In which cases can `r_norm` and `v_norm` be 0?

File: gmat/tests/build_frames_scripts.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+                    # Make a position vector who's norm is equal to the body's radius. Make a
+                    # unitary velocity vector. Units are in [m] and [m / s].
+                    rx, ry, rz = [float(k_r * r_i) for r_i in r_vec]
+                    vx, vy, vz = [float(k_v * v_i) for v_i in v_vec]
```
Comment (astrojuanlu) [inline gmat/tests/build_frames_scripts.py:72]: Why do we need the `float` here?

File: gmat/tests/build_frames_scripts.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+                        rx=rx,
+                        ry=ry,
+                        rz=rz,
+                        vx=vx,
+                        vy=vy,
+                        vz=vz,
+                    )
+
+                    with open(f"frames_tests/{filename}", "w") as filled_template_file:
+                        print(filled_template, file=filled_template_file)
```
Comment (astrojuanlu) [inline gmat/tests/build_frames_scripts.py:86]: Shouldn't we use something like `filled_template_file.write(filled_template)` instead? Not a fan of `print`. Or, alternatively, use https://jinja.palletsprojects.com/en/2.11.x/api/#jinja2.environment.TemplateStream.dump.

File: gmat/tests/template_frames.script [hunk 9]
```diff
@@ -0,0 +1,113 @@
+%----------------------------------------
```
Comment (astrojuanlu) [inline gmat/tests/template_frames.script:1]: Let's rename this to `template_frames.script.tpl`, so it's clear it's a template.

File: orekit/tests/validate_frames.py [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+    R = BODY_POLIASTRO.R.to(u.m).value
+
+    # Correction factor to normalize position and velocity vectors
+    k_r = R / r_norm if r_norm != 0 else 1.00
+    k_v = 1 / v_norm if v_norm != 0 else 1.00
+
+    # Make a position vector who's norm is equal to the body's radius. Make a
+    # unitary velocity vector. Units are in [m] and [m / s].
+    rx, ry, rz = [float(k_r * r_i) for r_i in r_vec]
+    vx, vy, vz = [float(k_v * v_i) for v_i in v_vec]
```
Comment (astrojuanlu) [inline orekit/tests/validate_frames.py:201]: Same as above

File: orekit/tests/validate_frames.py [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+        BODY_FIXED_FRAME_POLIASTRO,
+    ) = POLIASTRO_BODIES_AND_FRAMES[body_name]
+
+    # Compute for the norm of position and velocity vectors
+    r_norm, v_norm = [norm(vec) for vec in [r_vec, v_vec]]
+    R = BODY_POLIASTRO.R.to(u.m).value
+
+    # Correction factor to normalize position and velocity vectors
+    k_r = R / r_norm if r_norm != 0 else 1.00
+    k_v = 1 / v_norm if v_norm != 0 else 1.00
```
Comment (astrojuanlu) [inline orekit/tests/validate_frames.py:196]: Same as above

File: orekit/tests/validate_frames.py [hunk 12]
```diff
@@ -0,0 +1,307 @@
...
+@pytest.mark.parametrize("v_vec", V_SET)
+def validate_GCRF_to_ITRF(r_vec, v_vec):
+
+    # Compute for the norm of position and velocity vectors
+    r_norm, v_norm = [norm(vec) for vec in [r_vec, v_vec]]
+    R = Earth_poliastro.R.to(u.m).value
+
+    # Correction factor to normalize position and velocity vectors
+    k_r = R / r_norm if r_norm != 0 else 1.00
+    k_v = 1 / v_norm if v_norm != 0 else 1.00
```
Comment (astrojuanlu) [inline orekit/tests/validate_frames.py:268]: Same as above

File: orekit/tests/validate_frames.py [hunk 13]
```diff
@@ -0,0 +1,307 @@
...
+
+    # Make a position vector who's norm is equal to the body's radius. Make a
+    # unitary velocity vector. Units are in [m] and [m / s].
+    rx, ry, rz = [float(k_r * r_i) for r_i in r_vec]
+    vx, vy, vz = [float(k_v * v_i) for v_i in v_vec]
```
Comment (astrojuanlu) [inline orekit/tests/validate_frames.py:273]: Same as above

File: orekit/tests/validate_frames.py [hunk 14]
```diff
@@ -0,0 +1,307 @@
...
+# WGCCRE 2009 report while poliastro uses IAU WGCCRE 2015 one
+
+
+@pytest.mark.parametrize("r_vec", R_SET)
+@pytest.mark.parametrize("v_vec", V_SET)
+@pytest.mark.parametrize(
+    "body_name",
+    [
+        "Sun",
+        pytest.param("Mercury", marks=pytest.mark.xfail),
```
Comment (astrojuanlu) [inline orekit/tests/validate_frames.py:171]: Do we know why these tests fail? If not, we could add a comment like

```suggestion
        pytest.param("Mercury", marks=pytest.mark.xfail),  # reason unknown
```

**Review by astrojuanlu — CHANGES_REQUESTED** [review]
Did a first pass, this looks very good! Just added some style comments, and after that this is ready to go 🚀

File: gmat/tests/build_frames_scripts.py [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+from poliastro.bodies import Mercury as Mercury_poliastro
+from poliastro.bodies import Moon as Moon_poliastro
+from poliastro.bodies import Neptune as Neptune_poliastro
+from poliastro.bodies import Saturn as Saturn_poliastro
+from poliastro.bodies import Sun as Sun_poliastro
+from poliastro.bodies import Uranus as Uranus_poliastro
+from poliastro.bodies import Venus as Venus_poliastro
+
+# All interesting 3D directions
+R_SET, V_SET = [list(product([-1, 0, 1], repeat=3)) for _ in range(2)]
```
Comment (astrojuanlu) [inline gmat/tests/build_frames_scripts.py:17]: What about this?

```suggestion
R_SET = [vector for vector in product([-1, 0, 1], repeat=3) if vector != [0, 0, 0]]
V_SET = R_SET.copy()
```

Comment (jorgepiloto) [reply gmat/tests/build_frames_scripts.py:17]: Solved!

File: gmat/tests/build_frames_scripts.py (hunk 7 above)
Comment (jorgepiloto) [reply gmat/tests/build_frames_scripts.py:72]: Sorry, copy-paste error...

File: gmat/tests/build_frames_scripts.py (hunk 6 above)
Comment (jorgepiloto) [reply gmat/tests/build_frames_scripts.py:67]: Originates due to `product` function combination. Solved in previous suggestion 👍🏽

File: gmat/tests/build_frames_scripts.py (hunk 8 above)
Comment (jorgepiloto) [reply gmat/tests/build_frames_scripts.py:86]: Ups, sure! Solving... ⚙️

File: gmat/tests/template_frames.script (hunk 9 above)
Comment (jorgepiloto) [reply gmat/tests/template_frames.script:1]: Renaming... ✍🏽

File: orekit/tests/validate_frames.py (hunk 14 above)
Comment (jorgepiloto) [reply orekit/tests/validate_frames.py:171]: Sure! No problem. All of them are due to the fact that poliastro implements the WGCCRE 2015 report data while orekit is built on top WGCCRE 2009 one.

File: orekit/tests/validate_frames.py (hunk 10 above)
Comment (jorgepiloto) [reply orekit/tests/validate_frames.py:201]: In this case this is required as the orekit API is being used. Guess something related to the Java virtual machine or the wrapping nature of the package...

File: gmat/tests/build_frames_scripts.py (hunk 5 above)
Comment (jorgepiloto) [reply gmat/tests/build_frames_scripts.py:53]: Nope, you are right. Fixing indentation...

File: gmat/tests/build_frames_scripts.py (hunk 4 above)
Comment (jorgepiloto) [reply gmat/tests/build_frames_scripts.py]: Looks good!

File: gmat/tests/build_frames_scripts.py [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+                    vx=vx,
+                    vy=vy,
+                    vz=vz,
+                )
+
+                with open(f"frames_tests/{filename}", "w") as filled_template_file:
+                    # print(filled_template, file=filled_template_file)
+                    filled_template_file.write(filled_template)
+
+                # Close the filled template_file
```
Comment (astrojuanlu) [inline gmat/tests/build_frames_scripts.py:89]: No need to `.close()` the files, that's what the `with` blocks are for :)

**Review by astrojuanlu — APPROVED** [review]

**Comment by astrojuanlu** [comment]
Approved with one minor comment!

File: gmat/tests/build_frames_scripts.py (hunk 16 above)
Comment (jorgepiloto) [reply gmat/tests/build_frames_scripts.py]: Ups 😓

**Comment by jorgepiloto** [comment]
Merged! 😍

**Pull request merged**